src/dcc/Logon.hxx

```diff
@@ -47,40 +47,623 @@ namespace dcc
 /// Handler.
 class LogonHandlerModule
 {
+public:
+    /// @return the number of locomotives known. The locomotive IDs are
+    /// 0..num_locos() - 1.
+    unsigned num_locos();
+
+    /// @param loco_id a locomotive identifier
+    /// @return true if this is valid and belongs to a loco we know about.
+    bool is_valid_loco_id(unsigned loco_id);
+
+    /// Finds the storage cell for a locomotive and returns the flag byte for
+    /// it.
+    /// @param loco_id a valid locomotive ID.
+    /// @return the flag byte for this loco.
+    uint8_t &loco_flags(unsigned loco_id);
+
+    /// Retrieves the decoder unique ID.
+    /// @param loco_id the dense locomotive identifier.
+    /// @return the decoder unique ID (44 bit, LSb-aligned).
+    uint64_t loco_did(unsigned loco_id);
+
+    /// Creates a new locomotive by decoder ID, or looks up an existing
+    /// locomotive by decoder ID.
+    /// @param decoder_id 44-bit decoder ID (aligned to LSb).
+    /// @return locomotive ID for this cell.
+    unsigned create_or_lookup_loco(uint64_t decoder_id);
+
+    /// Runs the locomotive address policy. After the address policy is run,
+    /// the loco should have the ability to answer the assigned_address
+    /// question.
+    /// @param loco_id which locomotive this is
+    /// @param desired_address the S-9.2.1.1 encoded desired address for this
+    /// decoder.
+    void run_address_policy(unsigned loco_id, uint16_t desired_address);
+
+    /// @param loco_id
+    /// @return the address to be assigned to this locomotive. 14-bit.
+    uint16_t assigned_address(unsigned loco_id);
+
+    /// Flags for the logon handler module.
+    enum Flags
+    {
+        /// This decoder needs a get shortinfo command.
+        FLAG_NEEDS_GET_SHORTINFO = 0x01,
+        /// We sent a get shortinfo command.
+        FLAG_PENDING_GET_SHORTINFO = 0x02,
+
+        /// This decoder needs an assign command.
+        FLAG_NEEDS_ASSIGN = 0x04,
+        /// We sent an assign command
+        FLAG_PENDING_ASSIGN = 0x08,
 
+        /// 1 if we completed the address assignment.
+        FLAG_COMPLETE = 0x10,
+        /// 1 if we ended up in an error state for this loco.
+        FLAG_ERROR_STATE = 0x20,
+        /// 1 if we have asked for a re-try.
+        FLAG_PENDING_RETRY = 0x40,
+        /// This decoder needs a get shortinfo command.
```

Comment:
How is this different from FLAG_NEEDS_GET_SHORTINFO?

Reply:
this is a wrong comment. Fixed:
```
        /// This is a 1-bit pre-scaler on a shared 50 msec timer that controls
        /// the delay of re-tries. This makes a retry happen in 50 to 100 msec
        /// time from the original failed attempt.
        FLAG_PENDING_TICK = 0x80,
```